src/coreComponents/mesh/ParticleManager.hpp

```diff
@@ -0,0 +1,1514 @@
+   */
+  localIndex numParticleBlocks() const;
+
+  /**
+   * @brief This function is used to launch kernel function over all the particle regions with region type =
+   * ParticleRegionBase.
+   * @tparam LAMBDA type of the user-provided function
+   * @param lambda kernel function
+   */
+  template< typename REGIONTYPE = ParticleRegionBase, typename ... REGIONTYPES, typename LAMBDA >
+  void forParticleRegions( LAMBDA && lambda )
+  {
+    this->getGroup( groupKeyStruct::particleRegionsGroup() ).forSubGroups< REGIONTYPE, REGIONTYPES... >( std::forward< LAMBDA >( lambda ) );
+  }
+
+  /**
+   * @brief This const function is used to launch kernel function over all the particle regions with region type =
+   * ParticleRegionBase.
+   * @tparam LAMBDA type of the user-provided function
+   * @param lambda kernel function
+   */
+  template< typename REGIONTYPE = ParticleRegionBase, typename ... REGIONTYPES, typename LAMBDA >
+  void forParticleRegions( LAMBDA && lambda ) const
+  {
+    this->getGroup( groupKeyStruct::particleRegionsGroup() ).forSubGroups< REGIONTYPE, REGIONTYPES... >( std::forward< LAMBDA >( lambda ) );
+  }
+
+  /**
+   * @brief This function is used to launch kernel function over the target particle regions with region type =
+   * ParticleRegionBase.
+   * @tparam LOOKUP_CONTAINER type of container of names or indices
+   * @tparam LAMBDA type of the user-provided function
+   * @param targetRegions target particle region names or indices
+   * @param lambda kernel function
+   */
+  template< typename REGIONTYPE = ParticleRegionBase, typename ... REGIONTYPES, typename LOOKUP_CONTAINER, typename LAMBDA >
+  void forParticleRegions( LOOKUP_CONTAINER const & targetRegions, LAMBDA && lambda )
+  {
+    this->getGroup( groupKeyStruct::particleRegionsGroup() ).forSubGroups< REGIONTYPE, REGIONTYPES... >( targetRegions, std::forward< LAMBDA >( lambda ) );
+  }
+
+  /**
+   * @brief This const function is used to launch kernel function over the target particle regions with region type =
+   * ParticleRegionBase.
+   * @tparam LOOKUP_CONTAINER type of container of names or indices
+   * @tparam LAMBDA type of the user-provided function
+   * @param targetRegions target particle region names or indices
+   * @param lambda kernel function
+   */
+  template< typename REGIONTYPE = ParticleRegionBase, typename ... REGIONTYPES, typename LOOKUP_CONTAINER, typename LAMBDA >
+  void forParticleRegions( LOOKUP_CONTAINER const & targetRegions, LAMBDA && lambda ) const
+  {
+    this->getGroup( groupKeyStruct::particleRegionsGroup() ).forSubGroups< REGIONTYPE, REGIONTYPES... >( targetRegions, std::forward< LAMBDA >( lambda ) );
+  }
+
+  /**
+   * @brief This const function is used to launch kernel function over all the types of particle regions.
+   * @tparam LAMBDA type of the user-provided function
+   * @param lambda kernel function
+   */
+  template< typename LAMBDA >
+  void forParticleRegionsComplete( LAMBDA lambda ) const
```

Comment:
can you be a bit more strict on your visibilities?
It looks like `forParticleRegionsComplete` could be `private`?
And do you need all these flavors?